File: src/application/lead_service.rs
use crate::{
    application::error::ApplicationError,
    domain::{
        company::{Company, CompanyResponse},
        concern::{Concern, ConcernResponse},
        contact::{Contact, ContactResponse},
        lead::*,
        position::{Position, PositionResponse},
    },
};
use sqlx::PgPool;
use tracing::{info, error};

pub struct LeadService {
    pool: PgPool,
}

impl LeadService {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    fn calculate_maturity_score(&self, payload: &LeadPayload) -> i32 {
        let mut score = 0;
        if let Some(company_size) = payload.company_size {
            if company_size > 100 {
                score += 1;
            }
        }
        if let Some(estimated_users) = payload.estimated_users {
            if estimated_users > 50 {
                score += 1;
            }
        }
        if payload.concerns.len() > 2 {
            score += 1;
        }
        if let Some(job_title) = &payload.job_title {
            let title = job_title.to_lowercase();
            if title.contains("manager")
                || title.contains("director")
                || title.contains("cto")
                || title.contains("ceo")
            {
                score += 1;
            }
        }
        score.min(5)
    }

    pub async fn create_lead(
        &self,
        lead_payload: LeadPayload,
        altcha: String,
        visitor_id: Option<String>,
    ) -> Result<LeadResponse, ApplicationError> {
        info!("Starting create_lead transaction");
        let mut tx = self.pool.begin().await?;

        info!("Fetching contact");
        let contact = sqlx::query_as!(
            Contact,
            "SELECT * FROM contacts WHERE email = $1",
            &lead_payload.email
        )
        .fetch_optional(&mut *tx)
        .await?;

        let contact = if let Some(mut contact) = contact {
            info!("Updating existing contact");
            contact.name = lead_payload.name.clone();
            contact.phone = lead_payload.phone.as_ref().map(|s| s.to_string());
            contact.job_title = lead_payload.job_title.as_ref().map(|s| s.to_string());
            sqlx::query!(
                "UPDATE contacts SET name = $1, phone = $2, job_title = $3, updated_at = NOW() WHERE id = $4",
                contact.name,
                contact.phone,
                contact.job_title,
                contact.id
            )
            .execute(&mut *tx)
            .await?;
            contact
        } else {
            info!("Creating new contact");
            sqlx::query_as!(
                Contact,
                "INSERT INTO contacts (name, email, phone, job_title) VALUES ($1, $2, $3, $4) RETURNING *",
                lead_payload.name,
                lead_payload.email,
                lead_payload.phone,
                lead_payload.job_title
            )
            .fetch_one(&mut *tx)
            .await?
        };

        info!("Fetching company");
        let company = sqlx::query_as!(
            Company,
            "SELECT * FROM companies WHERE name = $1",
            &lead_payload.company_name
        )
        .fetch_optional(&mut *tx)
        .await?;

        let company = if let Some(mut company) = company {
            info!("Updating existing company");
            company.size = lead_payload.company_size;
            sqlx::query!(
                "UPDATE companies SET size = $1 WHERE id = $2",
                company.size,
                company.id
            )
            .execute(&mut *tx)
            .await?;
            company
        } else {
            info!("Creating new company");
            sqlx::query_as!(
                Company,
                "INSERT INTO companies (name, size) VALUES ($1, $2) RETURNING *",
                lead_payload.company_name,
                lead_payload.company_size
            )
            .fetch_one(&mut *tx)
            .await?
        };

        let maturity_score = self.calculate_maturity_score(&lead_payload);

        let pack_name = if lead_payload.concerns.iter().any(|c| c.contains("confiance")) {
            "confiance"
        } else if lead_payload.concerns.iter().any(|c| c.contains("croissance")) {
            "croissance"
        } else {
            "conformité"
        };
        info!("Fetching recommended pack");
        let recommended_pack = sqlx::query_as!(
            RecommendedPack,
            "SELECT * FROM recommended_packs WHERE name = $1",
            pack_name
        )
        .fetch_one(&mut *tx)
        .await?;

        info!("Fetching urgency");
        let urgency = sqlx::query_as!(
            LeadUrgency,
            "SELECT * FROM lead_urgencies WHERE name = $1",
            &lead_payload.urgency
        )
        .fetch_one(&mut *tx)
        .await?;

        info!("Fetching status");
        let status = sqlx::query_as!(
            LeadStatus,
            "SELECT * FROM lead_statuses WHERE name = 'nouveau'"
        )
        .fetch_one(&mut *tx)
        .await?;

        info!("Creating lead");
        let lead = sqlx::query_as!(
            Lead,
            r#"
            INSERT INTO leads (
                contact_id, company_id, problem_summary, estimated_users, urgency_id,
                status_id, maturity_score, recommended_pack_id, altcha_solution,
                fingerprint_visitor_id
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
            RETURNING *
            "#,
            contact.id,
            company.id,
            lead_payload.problem_summary,
            lead_payload.estimated_users,
            urgency.id,
            status.id,
            maturity_score,
            recommended_pack.id,
            altcha,
            visitor_id
        )
        .fetch_one(&mut *tx)
        .await.map_err(|e| {
            error!("Failed to create lead: {}", e);
            e
        })?;

        let mut positions = Vec::new();
        for position_title in &lead_payload.positions {
            info!("Fetching position: {}", position_title);
            let position = sqlx::query_as!(
                Position,
                "SELECT * FROM positions WHERE title = $1",
                position_title
            )
            .fetch_optional(&mut *tx)
            .await?;

            let position = if let Some(position) = position {
                position
            } else {
                info!("Creating new position: {}", position_title);
                sqlx::query_as!(
                    Position,
                    "INSERT INTO positions (title) VALUES ($1) RETURNING *",
                    position_title
                )
                .fetch_one(&mut *tx)
                .await?
            };

            info!("Linking position to lead");
            sqlx::query!(
                "INSERT INTO lead_positions (lead_id, position_id) VALUES ($1, $2)",
                lead.id,
                position.id
            )
            .execute(&mut *tx)
            .await?;
            positions.push(position);
        }

        let mut concerns = Vec::new();
        for concern_label in &lead_payload.concerns {
            info!("Fetching concern: {}", concern_label);
            let concern = sqlx::query_as!(
                Concern,
                "SELECT * FROM concerns WHERE label = $1",
                concern_label
            )
            .fetch_optional(&mut *tx)
            .await?;

            let concern = if let Some(concern) = concern {
                concern
            } else {
                info!("Creating new concern: {}", concern_label);
                sqlx::query_as!(
                    Concern,
                    "INSERT INTO concerns (label) VALUES ($1) RETURNING *",
                    concern_label
                )
                .fetch_one(&mut *tx)
                .await?
            };

            info!("Linking concern to lead");
            sqlx::query!(
                "INSERT INTO lead_concerns (lead_id, concern_id) VALUES ($1, $2)",
                lead.id,
                concern.id
            )
            .execute(&mut *tx)
            .await?;
            concerns.push(concern);
        }

        let potential_score = calculate_potential_score(&company, &urgency, &contact);

        info!("Committing transaction");
        tx.commit().await?;

        let response = LeadResponse {
            id: lead.id,
            submission_date: lead.submission_date,
            status: status.into(),
            urgency: urgency.into(),
            recommended_pack: Some(recommended_pack.into()),
            maturity_score: Some(lead.maturity_score.unwrap_or(0)),
            potential_score: Some(potential_score),
            estimated_users: lead.estimated_users,
            problem_summary: lead.problem_summary,
            contact: contact.into(),
            company: company.into(),
            positions: positions.into_iter().map(|p| p.into()).collect(),
            concerns: concerns.into_iter().map(|c| c.into()).collect(),
        };

        Ok(response)
    }
}

File: application/lead_service.py
"""Lead application service - refactored to use repository pattern."""

import logging
from datetime import datetime
from typing import List, Optional

from domain.repositories.lead_repository import LeadRepository
from domain.repositories.contact_repository import ContactRepository
from domain.repositories.company_repository import CompanyRepository
from domain.repositories.position_repository import PositionRepository
from domain.repositories.concern_repository import ConcernRepository
from domain.repositories.note_repository import NoteRepository
from domain.entities.lead import Lead, LeadStatus, LeadUrgency, RecommendedPack
from domain.entities.contact import Contact
from domain.entities.company import Company
from domain.entities.position import Position
from domain.entities.concern import Concern
from domain.entities.note import Note, NoteReason
from domain.services.lead_scoring_service import LeadScoringService
from domain.contact import LeadPayload, LeadUpdateRequest
# Still need ORM models for LeadPosition, LeadConcern, LeadModificationLog
from infrastructure.persistence.models import (
    LeadPositionModel as LeadPosition,
    LeadConcernModel as LeadConcern,
    LeadModificationLogModel as LeadModificationLog,
    LeadStatusModel as LeadStatusORM,
    LeadUrgencyModel as LeadUrgencyORM,
    RecommendedPackModel as RecommendedPackORM,
    LeadModel as LeadORM
)
from sqlalchemy.orm import Session

logger = logging.getLogger(__name__)


class LeadService:
    """Application service for lead operations - uses repository pattern."""

    def __init__(
        self,
        session: Session,  # Still needed for LeadPosition/LeadConcern/LeadModificationLog
        lead_repository: LeadRepository,
        contact_repository: ContactRepository,
        company_repository: CompanyRepository,
        position_repository: PositionRepository,
        concern_repository: ConcernRepository,
        note_repository: NoteRepository,
        scoring_service: LeadScoringService
    ):
        self._session = session
        self._lead_repo = lead_repository
        self._contact_repo = contact_repository
        self._company_repo = company_repository
        self._position_repo = position_repository
        self._concern_repo = concern_repository
        self._note_repo = note_repository
        self._scoring_service = scoring_service

    def _get_or_create_contact(
        self, name: str, email: str, phone: str, job_title: str, conscent: bool = False
    ) -> Contact:
        """Get or create a contact."""
        contact = self._contact_repo.find_by_email(email)
        if not contact:
            contact = Contact(
                id=None,
                name=name,
                email=email,
                phone=phone,
                job_title=job_title,
                conscent=conscent,
                created_at=datetime.now(),
                updated_at=datetime.now()
            )
            contact = self._contact_repo.save(contact)
        else:
            # Update contact info if it has changed
            contact.name = name
            contact.phone = phone
            contact.job_title = job_title
            contact.conscent = conscent
            contact.updated_at = datetime.now()
            contact = self._contact_repo.update(contact)

        return contact

    def _get_or_create_company(self, company_name: str, size: int) -> Company:
        """Get or create a company."""
        company = self._company_repo.find_by_name(company_name)
        if not company:
            company = Company(
                id=None,
                name=company_name,
                size=size
            )
            company = self._company_repo.save(company)
        else:
            company.size = size
            company = self._company_repo.update(company)

        return company

    def _get_or_create_position(self, position_title: str) -> Position:
        """Get or create a position."""
        position = self._position_repo.find_by_title(position_title)
        if not position:
            position = Position(
                id=None,
                title=position_title
            )
            position = self._position_repo.save(position)

        return position

    def _get_or_create_concern(self, concern_label: str) -> Concern:
        """Get or create a concern."""
        concern = self._concern_repo.find_by_label(concern_label)
        if not concern:
            concern = Concern(
                id=None,
                label=concern_label
            )
            concern = self._concern_repo.save(concern)

        return concern

    def create_lead(self, lead_payload: LeadPayload, altcha: str, visitor_id: str) -> LeadORM:
        """Create a new lead - returns ORM model for API compatibility."""
        try:
            # Get or create contact and company
            contact = self._get_or_create_contact(
                name=lead_payload.name,
                email=lead_payload.email,
                phone=lead_payload.phone,
                job_title=lead_payload.job_title,
                conscent=lead_payload.conscent
            )

            company = self._get_or_create_company(
                company_name=lead_payload.company_name,
                size=lead_payload.company_size
            )

            # Calculate maturity score using domain service
            maturity_score = self._scoring_service.calculate_maturity_score(lead_payload)

            # Get recommended pack using domain service
            pack_name = self._scoring_service.recommend_pack(lead_payload.concerns)
            recommended_pack_orm = self._session.query(RecommendedPackORM).filter_by(name=pack_name).first()
            if not recommended_pack_orm:
                logger.error(f"Recommended pack '{pack_name}' not found in the database.")
                recommended_pack_orm = self._session.query(RecommendedPackORM).filter_by(name='conformité').first()

            recommended_pack = RecommendedPack(
                id=recommended_pack_orm.id,
                name=recommended_pack_orm.name
            ) if recommended_pack_orm else None

            # Get urgency
            urgency_orm = self._session.query(LeadUrgencyORM).filter_by(name=lead_payload.urgency).first()
            if not urgency_orm:
                logger.warning(f"Urgency '{lead_payload.urgency}' not found, defaulting to 'moyen terme'.")
                urgency_orm = self._session.query(LeadUrgencyORM).filter_by(name='moyen terme').first()

            urgency = LeadUrgency(
                id=urgency_orm.id,
                name=urgency_orm.name
            ) if urgency_orm else None

            # Get status
            status_orm = self._session.query(LeadStatusORM).filter_by(name='nouveau').first()
            if not status_orm:
                logger.error("Default lead status 'nouveau' not found in the database.")
                raise ValueError("Initial lead status 'nouveau' is not configured in the system.")

            status = LeadStatus(
                id=status_orm.id,
                name=status_orm.name
            )

            # Get positions and concerns
            positions = [self._get_or_create_position(title) for title in lead_payload.positions]
            concerns = [self._get_or_create_concern(label) for label in lead_payload.concerns]

            # Create lead domain entity
            lead = Lead(
                id=None,
                submission_date=datetime.now(),
                estimated_users=lead_payload.estimated_users,
                problem_summary=lead_payload.problem_summary,
                maturity_score=maturity_score,
                altcha_solution=altcha,
                fingerprint_visitor_id=visitor_id,
                created_at=datetime.now(),
                updated_at=datetime.now(),
                contact=contact,
                company=company,
                status=status,
                urgency=urgency,
                recommended_pack=recommended_pack,
                positions=positions,
                concerns=concerns
            )

            # Save lead via repository
            saved_lead = self._lead_repo.save(lead)

            # Create LeadPosition and LeadConcern associations (still ORM-based)
            for position in positions:
                lead_position = LeadPosition(lead_id=saved_lead.id, position_id=position.id)
                self._session.add(lead_position)

            for concern in concerns:
                lead_concern = LeadConcern(lead_id=saved_lead.id, concern_id=concern.id)
                self._session.add(lead_concern)

            self._session.commit()

            # Return ORM model for API compatibility
            lead_orm = self._session.query(LeadORM).filter(LeadORM.id == saved_lead.id).first()
            return lead_orm

        except Exception as e:
            logger.exception("Error creating lead")
            raise e

    def get_all_leads(self) -> List[LeadORM]:
        """Get all leads - returns ORM models for API compatibility."""
        try:
            leads = self._lead_repo.find_all()
            # Convert back to ORM for API compatibility
            lead_ids = [lead.id for lead in leads]
            return self._session.query(LeadORM).filter(LeadORM.id.in_(lead_ids)).all()
        except Exception as e:
            logger.exception("Error getting all leads")
            raise e

    def get_lead_by_id(self, lead_id: int) -> Optional[LeadORM]:
        """Get lead by ID - returns ORM model for API compatibility."""
        try:
            lead = self._lead_repo.find_by_id(lead_id)
            if not lead:
                return None
            # Convert back to ORM for API compatibility
            return self._session.query(LeadORM).filter(LeadORM.id == lead.id).first()
        except Exception as e:
            logger.exception(f"Error getting lead by id {lead_id}")
            raise e

    def update_lead_notes(self, lead_id: int, notes: str) -> LeadORM:
        """Update lead notes - returns ORM model for API compatibility."""
        try:
            lead_orm = self._session.query(LeadORM).filter(LeadORM.id == lead_id).one_or_none()
            if lead_orm:
                # Get or create a "note interne" reason for internal notes
                reason = self._note_repo.find_reason_by_name("note interne")
                if not reason:
                    # Fallback to first available reason
                    reasons = self._note_repo.get_all_reasons()
                    if reasons:
                        reason = reasons[0]
                    else:
                        # Create default reason (still ORM-based for now)
                        from infrastructure.persistence.models import NoteReasonModel as NoteReasonORM
                        reason_orm = NoteReasonORM(name="note interne")
                        self._session.add(reason_orm)
                        self._session.commit()
                        self._session.refresh(reason_orm)
                        reason = NoteReason(id=reason_orm.id, name=reason_orm.name)

                # Create note domain entity
                note = Note(
                    id=None,
                    note=notes,
                    created_at=datetime.now(),
                    author_name="system",
                    lead_id=lead_id,
                    reason=reason
                )
                self._note_repo.save(note)

                self._session.refresh(lead_orm)

            return lead_orm
        except Exception as e:
            logger.exception(f"Error updating notes for lead {lead_id}")
            raise e

    def update_lead(self, lead_id: int, lead_update: LeadUpdateRequest) -> Optional[LeadORM]:
        """Update lead - returns ORM model for API compatibility."""
        lead_orm = self.get_lead_by_id(lead_id)
        if not lead_orm:
            return None

        if lead_orm.fingerprint_visitor_id != lead_update.visitorId or lead_orm.altcha_solution != lead_update.altcha:
            raise ValueError("Fingerprint or Altcha solution does not match.")

        for field, value in lead_update.model_dump(exclude_unset=True).items():
            if field in ["altcha", "visitorId"]:
                continue

            # Handle related fields
            if field in ["name", "email", "phone", "job_title", "conscent"]:
                old_value = getattr(lead_orm.contact, field)
                if old_value != value:
                    log_entry = LeadModificationLog(
                        lead_id=lead_id, field_name=field, old_value=str(old_value), new_value=str(value)
                    )
                    self._session.add(log_entry)
                    setattr(lead_orm.contact, field, value)
            elif field in ["company_name", "company_size"]:
                if field == "company_name":
                    old_value = lead_orm.company.name
                    if old_value != value:
                        log_entry = LeadModificationLog(
                            lead_id=lead_id, field_name=field, old_value=str(old_value), new_value=str(value)
                        )
                        self._session.add(log_entry)
                        lead_orm.company.name = value
                else:  # company_size
                    old_value = lead_orm.company.size
                    if old_value != value:
                        log_entry = LeadModificationLog(
                            lead_id=lead_id, field_name=field, old_value=str(old_value), new_value=str(value)
                        )
                        self._session.add(log_entry)
                        lead_orm.company.size = value
            elif field == "positions":
                # For simplicity, we replace all positions
                old_positions = [p.title for p in lead_orm.positions]
                log_entry = LeadModificationLog(
                    lead_id=lead_id, field_name=field, old_value=str(old_positions), new_value=str(value)
                )
                self._session.add(log_entry)

                self._session.query(LeadPosition).filter_by(lead_id=lead_orm.id).delete()
                for position_title in value:
                    position = self._get_or_create_position(position_title)
                    lead_position = LeadPosition(lead_id=lead_orm.id, position_id=position.id)
                    self._session.add(lead_position)
            elif field == "concerns":
                # For simplicity, we replace all concerns
                old_concerns = [c.label for c in lead_orm.concerns]
                log_entry = LeadModificationLog(
                    lead_id=lead_id, field_name=field, old_value=str(old_concerns), new_value=str(value)
                )
                self._session.add(log_entry)

                self._session.query(LeadConcern).filter_by(lead_id=lead_orm.id).delete()
                for concern_label in value:
                    concern = self._get_or_create_concern(concern_label)
                    lead_concern = LeadConcern(lead_id=lead_orm.id, concern_id=concern.id)
                    self._session.add(lead_concern)
            elif field == "urgency":
                # Handle urgency by looking up the LeadUrgency enum
                urgency = self._session.query(LeadUrgencyORM).filter_by(name=value).one_or_none()
                if not urgency:
                    raise ValueError(f"Invalid urgency value: {value}")
                old_value = lead_orm.urgency.name if lead_orm.urgency else None
                if old_value != value:
                    log_entry = LeadModificationLog(
                        lead_id=lead_id, field_name=field, old_value=str(old_value), new_value=str(value)
                    )
                    self._session.add(log_entry)
                    lead_orm.urgency_id = urgency.id
            else:
                old_value = getattr(lead_orm, field)
                if old_value != value:
                    log_entry = LeadModificationLog(
                        lead_id=lead_id, field_name=field, old_value=str(old_value), new_value=str(value)
                    )
                    self._session.add(log_entry)
                    setattr(lead_orm, field, value)

        self._session.commit()
        self._session.refresh(lead_orm)
        return lead_orm
